common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java

```diff
@@ -979,81 +1093,115 @@ private boolean ensureThreadStarted(int oldState) {
     }
 
     private void doStartThread() {
-        assert thread == null;
         executor.execute(new Runnable() {
             @Override
             public void run() {
+                processingLock.lock();
+                assert thread == null;
                 thread = Thread.currentThread();
                 if (interrupted) {
                     thread.interrupt();
+                    interrupted = false;
                 }
-
                 boolean success = false;
                 updateLastExecutionTime();
                 try {
-                    SingleThreadEventExecutor.this.run();
-                    success = true;
-                } catch (Throwable t) {
-                    logger.warn("Unexpected exception from an event executor: ", t);
-                } finally {
                     for (;;) {
-                        int oldState = state;
-                        if (oldState >= ST_SHUTTING_DOWN || STATE_UPDATER.compareAndSet(
-                                SingleThreadEventExecutor.this, oldState, ST_SHUTTING_DOWN)) {
-                            break;
+                        SingleThreadEventExecutor.this.run();
+                        success = true;
+
+                        if (state == ST_SUSPENDING && !hasTasks() && nextScheduledTaskDeadlineNanos() == -1) {
+                            if (!STATE_UPDATER.compareAndSet(SingleThreadEventExecutor.this,
+                                    ST_SUSPENDING, ST_SUSPENDED)) {
+                                // Try again
+                                continue;
+                            }
```

Comment:
I think there is a race where with the `private execute(Runnable, boolean)` method where we can end up with an orphaned task.
```
Event loop thread: state == ST_SUSPENDING && !hasTasks() etc -> get to the CAS operation and sleep.
Worker thread: execute(task, whatever) -> adds task -> startThread() -> state == ST_SUSPENDING -> return; // have a task in queue
event loop thread: wake and perform CAS -> state = ST_SUSPENDED and `hasTasks() == true` -> thread terminates
```
Presuming that's true, we need to double check the conditions that allow us to terminate after the CAS and if they fail we need to try to re-enter the `ST_STARTED` state.

Reply:
I think you are right...